### ISSUE TITLE
npm install fails with "ENOENT: no such file or directory" on .DELETE files

### ISSUE THREAD

**Issue by dkent600** [npm/npm#17444]
#### I'm opening this issue because:

  - npm is failing for some reason I don't understand

#### What's going wrong?

This is happening to me more often than not. Example:

> _npm ERR! enoent ENOENT: no such file or directory, rename 'C:\[...]\node_modules\rx' -> '[...]\node_modules\.rx.DELETE'_

It is a different file every time, sometimes doesn't happen at all, but when it does, always with [something].DELETE.

Turning off virus protection doesn't help.

Running `npm cache clean -f`, as administrator doesn't help

The problem has been getting steadily worse since a week or so ago, but seems somewhat random as to which file it will fail upon.

#### How can the CLI team reproduce the problem?

run `npm install`
wait for it to fail

npm-debug.log file:  https://gist.github.com/dkent600/9cf97717cc3dbd553bea6c1a116145a7

### supporting information:

 - `npm -v` prints:  v5.0.3
 - `node -v` prints: v8.1.2
 - `npm config get registry` prints: https://registry.npmjs.org/
 - Windows, OS X/macOS, or Linux?:  Windows 10
 - Network issues:
   - Geographic location where npm was run:
   - [ ] I use a proxy to connect to the npm registry.
   - [ ] I use a proxy to connect to the web.
   - [ ] I use a proxy when downloading Git repos.
   - [ ] I access the npm registry via a VPN
   - [ ] I don't use a proxy, but have limited or unreliable internet access.
 - Container:
   - [ ] I develop using Vagrant on Windows.
   - [ ] I develop using Vagrant on OS X or Linux.
   - [ ] I develop / deploy using Docker.
   - [ ] I deploy to a PaaS (Triton, Heroku).

(none of the above)

**Comment by garrettloy** [comment]
Also experiencing this.  Simply running npm install repeatedly eventually works for me....

![image](https://user-images.githubusercontent.com/7387847/27554859-b8889d24-5a64-11e7-9de1-fd27e023ec98.png)

In that screencap, the third 'node install' completed successfully.. no other commands were run. 

Have seen .rxjs.DELETE referenced multiple times, other missing .DELETE files are pretty sporadic

**Comment by dkent600** [comment]
Yes, same here.

**Comment by nerdybeast** [comment]
Same here

```
10305 silly finalize C:\projects\simple-vf-cli\node_modules\core-js
10306 verbose unlock done using C:\Users\xxxxxxxxx\AppData\Roaming\npm-cache\_locks\staging-0a8fbe58d39bfe29.lock for C:\projects\simple-vf-cli\node_modules\.staging
10307 verbose stack Error: ENOENT: no such file or directory, rename 'C:\projects\simple-vf-cli\node_modules\core-js' -> 'C:\projects\simple-vf-cli\node_modules\.core-js.DELETE'
10307 verbose stack     at Error (native)
10308 verbose cwd C:\projects\simple-vf-cli
10309 verbose Windows_NT 10.0.10586
10310 verbose argv "C:\\Program Files\\nodejs\\node.exe" "C:\\Users\\xxxxxxxxxx\\AppData\\Roaming\\npm\\node_modules\\npm\\bin\\npm-cli.js" "install" "pouchdb@latest"
10311 verbose node v6.9.3
10312 verbose npm  v5.0.3
10313 error path C:\projects\simple-vf-cli\node_modules\core-js
10314 error code ENOENT
10315 error errno -4058
10316 error syscall rename
10317 error enoent ENOENT: no such file or directory, rename 'C:\projects\simple-vf-cli\node_modules\core-js' -> 'C:\projects\simple-vf-cli\node_modules\.core-js.DELETE'
10318 error enoent This is related to npm not being able to find a file.
10319 verbose exit [ -4058, true ]
```

UPDATE:
I tries removing the node_modules folder and the package-lock.json file, still received this error.

**Comment by dkent600** [comment]
Same thing happens with `npm link`

**Comment by nerdybeast** [comment]
Yes, running `npm install` over and over eventually worked, seems like there is a race condition during the install command with npm5.

**Comment by dkent600** [comment]
it's getting worse and worse.  `npm` is becoming unusable.  looks like I'm going to have to try reinstalling it.

**Comment by dkent600** [comment]
After `npm` got to the point that it was unusable, just before reinstalling it, I had an idea:  I opened task manager (I'm using Windows) and found a bunch of instances of nodejs running.  I was able to shut them all down.  Then when I ran `npm install` again, which at this time was failing 100% of the time, and found that it was able to execute without any problem.

I'm going to try to figure out where these other instances of nodejs are coming from.  So far it looks like Visual Studio Code and Visual Studio 2017.

**Comment by MeirionHughes** [comment]
In addition to the .DELETE errors, I'm also getting rename errors - and it freezes for about 20 seconds ( `finalise:foo`) before issuing an error. example: 

```
npm ERR! path D:\Code\data.server\node_modules\.staging\ts-node-9bd1c32e
npm ERR! code EPERM
npm ERR! errno -4048
npm ERR! syscall rename
npm ERR! Error: EPERM: operation not permitted, rename 'D:\Code\data.server\node_modules\.staging\ts-node-9bd1c32e' -> 'D:\Code\data.server\node_modules\ts-node'
npm ERR!  { Error: EPERM: operation not permitted, rename 'D:\Code\data.server\node_modules\.staging\ts-node-9bd1c32e' -> 'D:\Code\data.server\node_modules\ts-node'
npm ERR!   cause:
npm ERR!    { Error: EPERM: operation not permitted, rename 'D:\Code\data.server\node_modules\.staging\ts-node-9bd1c32e' -> 'D:\Code\data.server\node_modules\ts-node'
npm ERR!      errno: -4048,
npm ERR!      code: 'EPERM',
npm ERR!      syscall: 'rename',
npm ERR!      path: 'D:\\Code\\data.server\\node_modules\\.staging\\ts-node-9bd1c32e',
npm ERR!      dest: 'D:\\Code\\data.server\\node_modules\\ts-node' },
npm ERR!   errno: -4048,
npm ERR!   code: 'EPERM',
npm ERR!   syscall: 'rename',
npm ERR!   path: 'D:\\Code\\data.server\\node_modules\\.staging\\ts-node-9bd1c32e',
npm ERR!   dest: 'D:\\Code\\data.server\\node_modules\\ts-node'
```

retrying a few times generally solves either error. I've noticed I'm getting the .DELETE error consistently with something like typescript or ts-node. dedupe race condition? 

Edit: I'm afraid this issue has led me to switch to yarn. 

Edit: possibly related: https://github.com/Microsoft/vscode/issues/29204

**Comment by peterbakonyi05** [comment]
Had same issue, thanks for @dkent600 for the suggested workaround (closing VS Code). At this point if I had the option to move to yarn, I would. Will do for my own projects...

**Comment by nerdybeast** [comment]
I've had many similar problems working with Yarn... I would caution with switching to Yarn, they don't handle compiled native C modules very well (for instance LevelDB which cannot be installed globally using Yarn for some reason).

**Comment by wfmmurali** [comment]
Hey guys ..I am still having the same issue(#8723 referred above) after following here and installing npm few different times . is Switching to Yarn is the only fix ??

a) Installed Npm 4 to 5 times . ran into the same issue
b) tried creating a new user account on my MacBook - nogo went into the same issue

**Comment by nerdybeast** [comment]
@wfmmurali by "Installed Npm 4 to 5 times" are you meaning you actually installed the npm cli 4 to 5 times or ran the `npm install` command that many times?  I eventually got it to work by running the `npm install` command several times.

**Comment by gust42** [comment]
have same problem, it works when i shutdown the Visual Studio Code thats open for that project

**Comment by MeirionHughes** [comment]
@wfmmurali are you using vscode?

**Comment by wfmmurali** [comment]
@nerdybeast Yes running the npm install command 4 to 5 times

@MeirionHughes no , not VSCODE. Java with Appium

**Comment by dkent600** [comment]
@wfmmurali  Check to see whether some other application has instances of node running in the background.  If so, try to terminate them and then retry your npm command.

**Comment by eimerreis** [comment]
@dkent600 thanks! That worked for me. Closing my editor and terminating all running node instances made it work

**Comment by janpio** [comment]
If you are using Visual Studio Code and have this problem, here is an issue about that you can track: https://github.com/Microsoft/vscode/issues/28593

**Comment by matthiasg** [comment]
Closing VS Code did solve the issue for me.

I would guess it might be a case not of a running node instance but of VS Code or similar tools watching the node_modules folder while it is being manipulated which can lead to locking issues.

Optimally npm should retry a few times before giving up or doing less manipulations like delete or rename (cannot delete/rename while a watcher is trying to read stats from the file)

**Comment by dkent600** [comment]
@matthiasg Two other tools have also been implicated besides Code (Visual Studio itself and another ["Java with Appium", whatever that is]).  

Have you ensured that there are no other instances of node running?

**Comment by matthiasg** [comment]
@dkent600 as I said, closing vs code did solve the issue. 

as for other apps also interfering I would assume they might be watching for filesystem changes as well and try to act upon each change right away. at least that would be a theory of mine.

**Comment by dkent600** [comment]
@matthiasg Oops, sorry, in haste I misread your comment.

**Comment by gucong3000** [comment]
Same to me.
node 8.1.2
npm 5.0.3
Windows 10.0.15063

```bash
E:\work\eclint (npm_dependencies) (eclint@2.2.0)
$ npm i
npm WARN deprecated node-uuid@1.4.8: Use uuid module instead
npm ERR! path E:\work\eclint\node_modules\tslint
npm ERR! code ENOENT
npm ERR! errno -4058
npm ERR! syscall rename
npm ERR! enoent ENOENT: no such file or directory, rename 'E:\work\eclint\node_modules\tslint' -> 'E:\work\eclint\node_modules\.tslint.DELETE'
npm ERR! enoent This is related to npm not being able to find a file.
npm ERR! enoent

npm ERR! A complete log of this run can be found in:
npm ERR!     C:\Users\cdyf\AppData\Roaming\npm-cache\_logs\2017-07-07T02_12_43_678Z-debug.log

E:\work\eclint (npm_dependencies) (eclint@2.2.0)
$ npm i
npm WARN deprecated node-uuid@1.4.8: Use uuid module instead
npm ERR! path E:\work\eclint\node_modules\tslint
npm ERR! code ENOENT
npm ERR! errno -4058
npm ERR! syscall rename
npm ERR! enoent ENOENT: no such file or directory, rename 'E:\work\eclint\node_modules\tslint' -> 'E:\work\eclint\node_modules\.tslint.DELETE'
npm ERR! enoent This is related to npm not being able to find a file.
npm ERR! enoent

npm ERR! A complete log of this run can be found in:
npm ERR!     C:\Users\cdyf\AppData\Roaming\npm-cache\_logs\2017-07-07T02_19_35_740Z-debug.log
```

**Comment by markwhitfeld** [comment]
Just as a thought, if there are many tools interfering with the current approach that npm is taking with the .DELETE and .STAGING folders, is there a different approach that won't have the same opportunities for tools to interfere with it?
It actually looks like the issue is occurring here:
[lib/install/action/finalize.js - line 52: function moveStagingToDestination... ](https://github.com/npm/npm/blob/19397ad523434656af3d3765e80e22d7e6305f48/lib/install/action/finalize.js#L52)
```javascript
function moveStagingToDestination () {
  return destinationIsClear()
    .then(actuallyMoveStaging)
    .catch(() => moveOldDestinationAway().then(actuallyMoveStaging))
}
```
The moveOldDestinationAway() call is throwing the error we see, but there is an underlying error being swallowed up (empty catch) that is being thrown by the actuallyMoveStaging call.
```javascript
function actuallyMoveStaging () {
  return move(extractedTo, pkg.realpath, moveOpts)
}
```

I suspect that the staging folder is being locked up by whichever tool is being used and then the move fails. The move is failing in our case because the source file is being locked up. By the action being performed in the catch I suspect that it was originally thought that the destination was getting in the way and needed removing... this is not true in our case.
The approach could be changed to storing the staging folders in a folder that the tools won't touch.
This could be outside of the project folder (and therefore the tool's reach - I think that this could apply to the DELETE, .staging and any other temp folders), or alternatively the 'node_modules/.staging' folder could be ignored explicitly by the tools.

Ignoring the node_modules/.staging folder may be a temporary workaround for this particular issue but I think that npm should adjust the approach to be more resilient to tools.

@zkat or @iarna (or whoever else from the team) what do you think?

**Comment by mjbvz** [comment]
For VSCode users that see this error while VSCode is running or while there are any lingering `Code-Helper` processes around: there was a bug in VSCode 1.13 that could cause the reported npm error. Here's an overview of the issue: https://github.com/Microsoft/vscode/issues/28593#issuecomment-313238852 

I have taken a fix in to VSCode 1.14 which should come out early next week. You can also try the fix in the current VSCode insiders build: https://code.visualstudio.com/insiders

**Comment by micah-williamson** [comment]
Closing VS Code worked for me.

**Comment by rogierschouten** [comment]
but closing all node-related processes isn't really a solution to this issue right? Please let't get this resolved because this is a blocking issue for us and we're still stuck using yarn

**Comment by ndeen86** [comment]
same like the above for me.
npm -v 5.0.3
node -v 6.10.1

npm WARN learning@1.0.0 No repository field.
npm WARN The package css-loader is included as both a dev and production dependency.
npm WARN The package style-loader is included as both a dev and production dependency.

npm ERR! path C:\Deen\Research Lab\WebPackLab\Learning\node_modules\tslint
npm ERR! code ENOENT
npm ERR! errno -4058
npm ERR! syscall rename
npm ERR! enoent ENOENT: no such file or directory, rename '\Learning\node_modules\tslint' -> '\node_modules\.tslint.DELETE'
npm ERR! enoent This is related to npm not being able to find a file.
npm ERR! enoent

npm ERR! A complete log of this run can be found in:
npm ERR!     \AppData\Roaming\npm-cache\_logs\2017-07-12T21_20_44_975Z-debug.log

**After I closed my VS Code it Worked!**

**Comment by janpio** [comment]
VSCode 1.14 is actually out, so updating should fix this permanently. Worked for me.

**Comment by ifrim** [comment]
I had the same problem described in this thread after upgrading my node install to 8.2.1 (npm 5.3.0). I don't use VSCode but I tried closing my editor and other applications to see if that works, it didn't.

Then I tried clearing the npm cache but the following message was printed to the console:
"As of npm@5, the npm cache self-heals from corruption issues and data extracted from the cache is guaranteed to be valid. If you want to make sure everything is consistent, use 'npm cache verify' instead."

So then I ran "npm cache verify" and one of the output lines was something like "Corrupted content removed: 36". After that everything worked, with my editor and any other application opened.

(I also had to update node-sass from version 4.3.0 to 4.5.3 but it seems unrelated to the issue)

**Comment by kicferksoftwire** [comment]
Had the same issue, closing Visual Studio Code fixed the problem.

**Comment by pacific21** [comment]
Yup, closing Visual Studio fixed the problem.

**Comment by quanpower** [comment]
antd-admin git:(master) ✗ npm install
npm WARN checkPermissions Missing write access to /home/quanpower/antd-admin/node_modules/_read-pkg-up@2.0.0@read-pkg-up/node_modules/find-up
npm ERR! path /home/quanpower/antd-admin/node_modules/_read-pkg-up@2.0.0@read-pkg-up/node_modules/find-up
npm ERR! code ENOENT
npm ERR! errno -2
npm ERR! syscall access
npm ERR! enoent ENOENT: no such file or directory, access '/home/quanpower/antd-admin/node_modules/_read-pkg-up@2.0.0@read-pkg-up/node_modules/find-up'
npm ERR! enoent This is related to npm not being able to find a file.
npm ERR! enoent 

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/quanpower/.npm/_logs/2017-08-07T09_34_38_562Z-debug.log


A same error!!and i have tested three versions,but got the same error! SO SAD~~~~

**Comment by stiekel** [comment]
I got the same rename error on CentOS 7, Node 8.2.1, NPM 5.3.0, after I killed all the Node processes, It's still can NOT install.

**Comment by natesilva** [comment]
Same error on macOS 10.12.6, Node 8.3.0, NPM 5.3.0. It often, but not always is referring to `.DELETE`. It does always report `ENOENT: no such file or directory, rename…`

There are no other Node processes running. I closed Visual Studio Code, still seeing ENOENT.

**Workaround:** Our standard workaround for this and all other NPM 5.x problems is:

* Delete `package-lock.json` and run NPM again.

And never commit the !@#$% infernal `package-lock.json` to Git, it causes way more problems than it solves.

**Comment by npHavens** [comment]
Same issue for me. Using nvm to switch from node v8.2.1 to node v6.11.2 then running npm install again seems to work for me.

**Comment by nickmason** [comment]
@natesilva Deleting `package-lock.json` and reinstalling is the only thing that worked for me as well, for the exact same rename error. macOS 10.12.6, Node 8.3.0, NPM 5.3.0.

**Comment by oneezy** [comment]
where is this `package-lock.json` file you guys are referring to? I don't have any other node instances running or VS code editor open.. I am however running [Bash On Windows](https://github.com/Microsoft/BashOnWindows) (the newest version w/ Interop), so this could be related to my issue.

**Comment by Butterfly** [comment]
@oneezy Do you have a tool to search your computer files? That's how I found mine. I'm using Debian Linux (Buster). I first run `updatedb` to register all the latest files and directories gained via recent installs. After that I search using `locate [file-name]`. There were several (rational) instances in different locations in my use case.

One place where it changes failure into success is those instances where we should be doing our work from within the directory containing it. I learned that lesson firsthand. :smile:

**Comment by seancheung** [comment]
Same here with sqlite3 and bcrypt modules  installed.
Tested with node 6/7/8 with npm 5.3 on Window 10, Ubuntu 16 and MacOS Sierra.
Closing VSCode or terminating node process didn't work for me.
However, removing lock file worked. I guess the lock file has some issues with native modules.
@nickmason @natesilva

**Comment by seancheung** [comment]
Also I notice that for cross-platform collaboration, the lock file seems redundant since it varies on different platforms.

**Comment by IstoraMandiri** [comment]
Also having this issue with atom editor on OSX; `npm ERR! enoent ENOENT: no such file or directory, rename....`. 

Closing atom allows `npm install` to work again, but it also seems to corrupt package-lock.json, so I have to nuke it to properly install.

**Comment by zelibobla** [comment]
Had the same issue. Removing node_modules folder and installing all the staff again helped.

**Comment by PatonLewis** [comment]
Had the same issue on Ubuntu with node 7.10.0, so this is not a VS problem. Removing the build products folder and re-trying did not fix the problem, but changing the order in which my node packages were built did fix the problem.

**Comment by cxbrgr** [comment]
Ran into the same issue with:
angular/cli: 1.4.4
node: 8.6.0
os:Windows10  x64

tried without success: 
executing `npm install` multiple times
closing visual studio code
restarting computer (as mentioned maybe files are locked)

finally able to install after:
`npm cache clean --force`

**Comment by jmmaranan** [comment]
Manually removing any of the {package}.DELETE in ~/AppData/Roaming/npm/node_modules after the initial `npm i -g {package}` and retrying `npm i -g {package}` works.

Windows 7
node 6.11.3
npm 3.10.10

All accounts used to test have Admin - but setting this privilege on the folder or using a console to Run As Admin doesn't work.

**Comment by iarna** [comment]
The first error in this thread is an ENOENT on 5.0.3. It describes a class of error that was fixed in 5.2.0, and if you're seeing exactly that I recommend you update. (The latest version, as of this writing, is 5.5.1)

Second the EPERM error from @MeirionHughes is entirely unrelated. (Adding unrelated issues to an issue thread makes things much harder to manage. In the future, please open a new issue.) It's actually unlike any EPERM we've seen previously as it's erroring when we go to move something from staging into its final destination. The staging location only exists for a few seconds. EPERM is issued (on Windows only) when something has a file in the directory open. As folks have linked this to a VSCode issue I assume it's scanning files as they appear in the directory which is then causing the conflict. 

I would love some suggestions on how to work around this kinds of Windows-only file contention issues, as we've kind of gone in circles on this for years. One option would be to make [move-concurrently](https://github.com/npm/move-concurrently/blob/latest/move.js#L75) fall back to it's copy + delete on Windows and EPERM. That might work, or it might just get us the same error when the delete happens. Another option would be to sleep some amount of time before retrying the rename. (We use `graceful-fs` for renames, which should be retrying already when we see EPERMs on Windows.)

**Comment by jmmaranan** [comment]
@iarna - I did some trial and error on graceful-fs/polyfill.js at line 84 as these next few set of lines barfs often at me. Closing VSCode does seem to work, but what if it is left open, what I think is to have `fs.rename` wait it out for a bit more. The default wait was 1 sec and 1 try. I edited and tried some slower params, 5 secs and trying max 10 times during any `npm i -g {package}`. Note, this same code block seems to be called during `npm uninstall -g {package}`:
```
  // on Windows, A/V software can lock the directory, causing this
  // to fail with an EACCES or EPERM if the directory contains newly
  // created files.  Try again on failure, for up to 1 second.
  if (process.platform === "win32") {
    var retries = 1
    var maxRetries = 10
    var waitTime = 5000 // milliseconds
    fs.rename = (function (fs$rename) { return function (from, to, cb) {
      var start = Date.now()
      fs$rename(from, to, function CB (er) {
        if (er
            && (er.code === "EACCES" || er.code === "EPERM")
            && Date.now() - start < 3000 && retries <= maxRetries) {
          console.log(`Error encountered with try ${retries}, retry in ${waitTime/1000} secs`)
          retries = retries + 1
          return fs$rename(from, to, CB)
        }
        if (cb) cb(er)
      })
    }})(fs.rename)
  }
```
It does seem to work for me. But I don't know if normal npm user would edit polyfill.js. Also, issue seems to be reproducible when the machine is in a lot of stress or too many apps open (VSCode is closed).

Update 1: Should have looked into `graceful-fs` first but I think this patch is the answer: https://github.com/isaacs/node-graceful-fs/pull/119

Update 2: I tried @mekwall's patch and it seems to work for me. Even with VSCode open.

Update 3: @mekwall - I did encounter the `.DELETE` issue on termux for Android (note 4). Just an FYI, not saying you should re-patch for non-win32/x64 :)

**Comment by Bombdog** [comment]
If it helps anyone I got this "ENOENT: no such file or directory, rename" type error under vagrant/vmware shared folders. Turned off the VM and hey presto the same command runs fine on the host filesystem

**Comment by jacec** [comment]
As with @Bombdog - I'm also seeing this problem. I'm relatively certain it's a race condition. I had a VMware fusion share between my mac (host) and an Ubuntu server (VM). running ```npm install``` on the VM would invariably fail with ```ENOENT: no such file or directory, rename ...``` and yet after the failure, I actually see the ```.DELETE``` file in question. If I copied the code from the share to a local directory in the VM (thus removing the latency of read/write from the shared disk) the problem was resolved.

**Comment by devcircus** [comment]
@oneezy I see you mentioned you were having this issue using Bash on Windows and not using VS Code. Did you figure this out? This is the only thing keeping me from using Ubuntu on Windows for my dev environment.

**Comment by afspear** [comment]
I too am experiencing this using Bash on Windows (WSL).  Using nvm I have experimented with several different versions of npm and node 6.x, 8.x, and 9.x. All of them give the same error:
```
$npm install -g @angular/cli
...
npm ERR!  { Error: EACCES: permission denied, rename '/home/afspear/.nvm/versions/node/v9.1.0/lib/node_modules/@angular/cli' -> '/home/afspear/.nvm/versions/node/v9.1.0/lib/node_modules/@angular/.cli.DELETE'
npm ERR!   cause:
npm ERR!    { Error: EACCES: permission denied, rename '/home/afspear/.nvm/versions/node/v9.1.0/lib/node_modules/@angular/cli' -> '/home/afspear/.nvm/versions/node/v9.1.0/lib/node_modules/@angular/.cli.DELETE'
npm ERR!      errno: -13,
npm ERR!      code: 'EACCES',
npm ERR!      syscall: 'rename',
npm ERR!      path: '/home/afspear/.nvm/versions/node/v9.1.0/lib/node_modules/@angular/cli',
npm ERR!      dest: '/home/afspear/.nvm/versions/node/v9.1.0/lib/node_modules/@angular/.cli.DELETE' },
```

**Comment by romiohasan** [comment]
Just had this similar problem on Windows 10 and Npm 5+. 
Fixed it by opening the cmd prompt in Administrative mode. Then npm install went just fine!

**Comment by devcircus** [comment]
Turns out my issue was our company's firewall and only c: drive was affected. Keeping my projects on another drive fixed the issue.

**Comment by alexleen** [comment]
Also having this issue with:
`npm -v` 5.5.1
`node -v` v8.9.1
Windows 10
when installing the Angular CLI:
`npm install -g @angular/cli`

I was able to workaround this issue in a similar way to @dkent600 - except I had to kill extra Node.js processes immediately after running `npm install`. It appears that 3 processes were spawned, 2 of which have 0% processor utilization. After killing these two processes the install worked.
**Edit:**
I am not running VS Code. Killing the extra processes doesn't always work. Deleting package-lock.json seems to work in this case.

**Comment by markwhitfeld** [comment]
@zkat @iarna or any other team members... 
Did anybody investigate the potential source of the issue that I described here in July?https://github.com/npm/npm/issues/17444#issuecomment-313670843 
I really do think that a change in approach could solve the issue.

**Comment by herrmanno** [comment]
I'm using npm@5.5.1 and still encounter this kind of error.

```
ENOENT: no such file or directory, rename '/Users/oliverherrmann/Dev/quiz-web/node_modules/npm/node_modules/ansicolors' -> '/Users/oliverherrmann/Dev/quiz-web/node_modules/npm/node_modules/.ansicolors.DELETE'
```
The only thing that fixed the problem in my case was to manually remove directories of *old* dependencies below my node_modules folder.

`rm -rf node_modules/npm` did fix my problem.

**Comment by anil1712** [comment]
Still getting the same
`$ npm i
npm WARN C:\Users\myuser\Projects\project1\node_modules\@types\react-test-renderer is not a child of C:\Users\myuser\Projects\project2
npm ERR! path C:\Users\myuser\Projects\project2\node_modules\.staging\@types\react-test-renderer-24738ad5
npm ERR! code ENOENT
npm ERR! errno -4058
npm ERR! syscall rename
npm ERR! enoent ENOENT: no such file or directory, rename 'C:\Users\myuser\Projects\project2\node_modules\.staging\@types\react-test-renderer-24738ad5' -> 'C:\Users\myuser\Projects\project1\node_modules\@types\react-test-renderer'
npm ERR! enoent This is related to npm not being able to find a file.
npm ERR! enoent

npm ERR! A complete log of this run can be found in:
npm ERR!     C:\Users\myuser\AppData\Roaming\npm-cache\_logs\2017-12-21T11_52_04_398Z-debug.log`

Any solution?

**Comment by natesilva** [comment]
For those on Windows and still having problems, try temporarily disabling your anti-virus software. I don’t think that is the cause of every case of this problem, but it did solve the issue for some of our users. And the antivirus vendor provided our IT people with a fix.

**Comment by wonder95** [comment]
Just to add my $.02, I was running into this issue with node 9.2.1 and npm 5.5.1 on Mac 10.12.16. Repeated running of `npm install` did not fix it. Based on @natesilva's [comment](https://github.com/npm/npm/issues/17444#issuecomment-321851320), I deleted my `package-lock.json` file, and the install ran fine without the error.

**Comment by jdurand** [comment]
This happens on macOS with any version of node or NPM. Rage deleting all or some node_modules fixes this temporarily, but it comes back intermittently.

**Comment by myke11j** [comment]
After trying everything from 

1. closing vs code
2. restarting system
3. to closing node instances

Removing package.lock file worked.

**Comment by nickserv** [comment]
I'm on macOS, not using VSCode. I tried `killall node`, removing `node_modules` and `npm cache clear --force`, but I ended up having to remove and rebuild my `package-lock.json`.

**Comment by wongjn** [comment]
On Windows 10 1709, I disabled Webroot (Anti-virus program) real time protection for it to work.

**Comment by todd** [comment]
Just ran into this myself - removing `package-lock.json` fixed it, but that really seems like a suboptimal fix for this problem.

**Comment by cornillemichiel** [comment]
@todd same here! node latest npm latest, vscode latest, only deleting package-lock.json fixed ENOENT issue.

**Comment by pooja51dr** [comment]
**# Solved my problem :-** 
Go to  folder where package-lock.json file located,delete this file and go to command prompt change directory wherever you have package.json file ,
 Then run command npm start 
**(eg. C:\Users\poojar>d:
**D:\MVC sample project\Angular2Demo\Angular2Demo>npm start)****  

It will work definitely :-)

**Comment by jgustrow** [comment]
FYI - It definitely did not work in my case. I got two packages to install, and the second only after a couple of tries. Non of the clunky workarounds I'm seeing here and elsewhere do any good. I have the latest versions of everything on Windows 7/10.

Npm needs to be fixed because at this point I can't get any new packages to install even after repeated attempts, removing the lock file, messing with file permissions in AppData, etc. I'm a node.js newbie, but not new to developing. I'm surprised that I'm seeing posts around the web going back four years for this problem, and there is still no fix. I would think this is the kind of bug that would get immediate attention. At any rate, my little project is dead in the water until it's fixed.

**Comment by arizzitano** [comment]
I've got a team of developers all hitting this issue after upgrading from npm 3 to npm 5.5.1. We are running:
- Docker containers running Ubuntu 16.04.2
- The container mounts the project directory from the host machine

The "close the editor" hack appears to have worked a handful of times, but not reliably. `npm i` seems to work from the host machine, but this is not a tenable workaround as we build these container images regularly in CI.

**Comment by TehShrike** [comment]
I'm experiencing this issue with the current latest version of npm (5.6.0) and node 9.3.0 on macOS.

I do not have VSCode installed.

**Comment by jnoksoko** [comment]
We are having this same problem in the CI system with ubuntu containers, even with npm 5.6.0. The problem also exists on developers machines, but locally it can be solved easily by deleting the package-lock file and reinstalling. We would like to be able to push the package-lock files to the git repositories to have stable releases, but seems like we need to stop using them before this issue has been fixed. 

Node: 8.9.4
OS: macOS and ubuntu (16.04)
NPM: 5.5.1 and 5.6

**Comment by StefanChristou** [comment]
I was having the same issue with `npm install` returning:

```
npm ERR! path /Users/*&%^&$/node_modules/npm/node_modules/dezalgo
npm ERR! code ENOENT
npm ERR! errno -2
npm ERR! syscall rename
npm ERR! enoent ENOENT: no such file or directory, rename '/Users/*&%^&$/node_modules/npm/node_modules/dezalgo' -> '/Users/*&%^&$/node_modules/npm/node_modules/.dezalgo.DELETE'
npm ERR! enoent This is related to npm not being able to find a file.
npm ERR! enoent

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/aName/.npm/_logs/2018-01-19T09_38_19_248Z-debug.log
```

**I resolved this issue by deleting package-lock.json**

**Comment by blinkbat** [comment]
deleting package-lock.json was the only solution for me as well. pain in the butt.

EDIT: win10 using subl and gitbash

**Comment by arizzitano** [comment]
My team piled onto this issue and was able to 1) discover some potentially relevant details, and 2) resolve it for our use case (running within Docker). The `ENOENT` occurs during the binary symlinking step of `npm install`. If you install with `--no-bin-links` (or add `bin-links = false` to your `.npmrc`) the installation may succeed, but any tasks that rely on these binaries will fail (npm scripts, etc). If your `package-lock.json` is corrupt in any way (especially if it has been generated on a filesystem other than the filesystem you're installing on) you're at greater danger of hitting the `ENOENT`. We had issues with `fsevents` being bundled into our `package-lock.json` when it was generated on MacOS, but installed within a container running Ubuntu.

Now, for our specific use case. All our containers run Ubuntu Xenial and we began seeing this issue after a recent upgrade to Node 8/npm 5. Interestingly, the problem did *not* occur for any developers on Linux systems. Only developers on MacOS were affected. It's unclear whether npm or Docker for Mac are to blame, but I can say with certainty that this issue did not occur for us before the npm 5 upgrade.

**To anyone experiencing this issue within Docker containers** -- we determined the problem:
- is filesystem-related,
- only occurs on MacOS systems,
- only occurs with `:cached` or `:delegated` mounts of the project folder
- does not recur consistently so may point to a race condition

We were able to resolve it by mounting `node_modules` as a separate volume that only exists within the container and is **not** mounted into the host at all. An alternative solution involves forcing a `:consistent` mount of the project folder into the container, but this solution is less performant so we stuck with the first solution.

Further detail is available in this ticket https://openedx.atlassian.net/browse/FEDX-500.

**Comment by quinnliu** [comment]
deleting `package-lock.json` then reinstalling works for me.

macOS 10.13.1
no visual studio
`node -v` = v8.9.3 
`npm -v` = 5.6.0

**Comment by alundiak** [comment]
I'm:
- MacOS 10.13.2
- no visual studio (Sublime Text 3 only, and closed)
- node 8.9.4 (I've just upgraded recently. MacOS not yet restarted)
- npm 5.6.0

My original `package.json` has many deps. But looks like `materialize-css` causes this `dezalgo rename problem` for me (under the hood it's jquery@3.3.0 issue):

```
  "dependencies": {
    "materialize-css": "^0.100.2"
  }
```

Note: I don't have any errors, when install first or second time, manually: `npm install materialize-css`. But using `npm install`, meaning I rely on package.json causes then the error.

**Yes**, for this use case, **when I delete** `package-lock.json` it helps, and then every `npm install` commands go **OK**.

**BUT**.

Here is my mostly realistic `package.json`:
```
  "dependencies": {
    "body-parser": "^1.18.2",
    "cookie-parser": "^1.4.3",
    "express": "^4.16.2",
    "materialize-css": "^0.100.2",
    "mongodb": "^3.0.1",
    "requirejs": "^2.3.5",
    "shelljs": "^0.7.8",
    "uuid": "^3.1.0"
  }
```


And before running npm install, I do "cleanMagic"
- `npm cache clean --force`
- `rm -rf node_modules`
- `rm package-lock.json`.

And guess what? Again: first `npm install` is OK, but second run bring me the error again.

```
npm ERR! path /Users/alund/prj/decision-making-conditions/node_modules/npm/node_modules/dezalgo
npm ERR! code ENOENT
npm ERR! errno -2
npm ERR! syscall rename
npm ERR! enoent ENOENT: no such file or directory, rename '/Users/alund/prj/decision-making-conditions/node_modules/npm/node_modules/dezalgo' -> '/Users/alund/prj/decision-making-conditions/node_modules/npm/node_modules/.dezalgo.DELETE'
npm ERR! enoent This is related to npm not being able to find a file.
npm ERR! enoent 

```

And no matter how many times I repeat

When clean install, then all works at first time. Here is my full `npm ls`
[npm_ls_after_first_npm_install.txt](https://github.com/npm/npm/files/1648779/npm_ls_after_first_npm_install.txt)

What I see, that nested npm version is 4.4.1 (which is required by JQuery 3.3.0), and I suspect it's related to versions misconfiguraion inside of ANY module, which use npm as dep.

Looks like they already have the issue reported: https://github.com/jquery/jquery/issues/3947


**And btw, what helped me** - upgrade materialize-css to `1.0.0-alpha.3` which uses jquery `3.2.1` and when installed, it installs properly. Old materialize-css : 
```
  "dependencies": {
    "hammerjs": "^2.0.8",
    "jquery": "^3.0.0 || ^2.1.4"
  },
```
An now, because new/latest/alpha materialize-css clear `"dependencies": {},` and uses only `"devDependencies"`, so it helps 3rd-party consumers projects like my to NOT have issues with dependencies installed locally.

**UPD**. JQuery released new version `3.3.1` which fixes above issue for me, with materialize-css.
I installed again materialize-css@latest and now `npm ls` is:
```
├─┬ materialize-css@0.100.2
│ ├── hammerjs@2.0.8
│ └── jquery@3.3.1
```

**Comment by natesilva** [comment]
@alundiak The issue you were having was due to `jquery@3.3.0` (a dependency of `materialize-css`). With `jquery@3.3.1`, published a few hours ago, the problem goes away. See https://github.com/jquery/jquery/issues/3947.

I’m not an expert on `package-lock.json` but I think you may need to remove it, otherwise it may insist on installing the broken `3.3.0` again. From that point forward you would be okay and should not have to remove `package-lock.json` again.

**Comment by alundiak** [comment]
@natesilva Yes, I know, and I updated my comment. Btw, I did not removed `package-lock.json`. But I did `rm -rf node_modules`. So for me, lock file looks not important.

**Comment by anandmahajan267** [comment]
Hi all,
just delete package-lock.json file and then install package what you want all will works :-)

**Comment by tylerjwatson** [comment]
Count me in to this issue.  

```bash
 602  npm i
  603  tsc
  604* npm i
  605* rm -rf node_modules
  606* npm i
  607* ../DoraToolbox
  608* ls
  609* rm -rf node_modules
  610* npm i
  611* tsc
  612* git pull
  613* tsc
  614* popd
  615* rm -rf node_modules
  616* npm i
  617* yarn install
  618* sudo dnf install yarn
  619* npm run build
  620* npm rm webpack awesome-typescript-loader source-map-loader
  621* npm i
  622* rm node_modules
  623* rm -rf node_modules
  624* rm package.json
  625* git checkout -- package.json
  626* rm package-lock.json
  627* npm i
  628* npm rm --save uglifyjs-webpack-plugin
  629* npm rm react-router-dom --save
  630* tsc
  631* npm rm react--dom --save
  632* npm rm react-dom --save
  633* tsc
  634* ../project-dora
  635* npm i
  636* npm i --save-dev cheerio enzyme
  637* ../DoraToolbox
  638* ls
  639* npm i
  640* npm i --save-dev react react-dom enzyme cheerio
  641* npm i
  642* tsc
  643* ../project-dora
  644* npm i
  645  ~/src
  646  cd DoraToolbox
  647  npm i
  648  ../project-dora
  649  npm i
  650  rm -rf node_modules
  651  npm i
  652  rm -rf node_modules/ package-lock.json
  653  npm i
  654  npm i --save-dev
  655  npm cache clear --force
  656  npm i --save-dev
  657  npm i
  658  ls
  659  rm -rf node_modules
  660  npm i
```

This `history` export is basically what it took me to resolve this issue.  I can speclate on what actually worked in the end though, in this order:

1. Delete node_modules
2. `npm cache clear --force`
3. Removing `package-lock.json`
4. `npm install`

The insane thing is though, that the next person who downloads my replaced `package-lock.json` file will run into the same problem if they clear `node_modules` and attempt a full reinstall:  Having local dependencies amplifies this issue to a common occurance.  npm@5 is bonkers.

**Comment by marcuscraske** [comment]
Mac / node 5.6.0, encountered this issue when upgrading jquery from 3.2.1 to 3.3.0.

@natesilva 3.3.1 does fix this issue, cheers!

**Comment by gyh000** [comment]
https://stackoverflow.com/questions/48161174/npm-install-gives-enoent-errno-2-re-missing-dezalgo-module
I'm from this question then I delete the package-lock.json file and it's work

**Comment by adamlubek** [comment]
What's the progress on this? 

We keep on getting:

```
npm ERR! path \node_modules\npm\node_modules\dezalgo
npm ERR! code ENOENT
npm ERR! errno -4058
npm ERR! syscall rename
npm ERR! enoent ENOENT: no such file or directory, rename '\node_modules\npm\node_modules\dezalgo' -> '\node_modules\npm\node_modules\.dezalgo.DELETE'
npm ERR! enoent This is related to npm not being able to find a file.
npm ERR! enoent

```

We're using:
```
node -v
v8.9.4
npm -v
5.6.0
```

The only thing that works consistently is deleting package-lock.json to fix this but we can't keep on doing it as it contradicts purpose of package-lock.json...

**Comment by jaylinski** [comment]
@adamlubek can you reproduce this issue? or does it occur randomly?

If you can reproduce it, please share your package.json, lock-file and environment, so I can try and debug it.

**Comment by adamlubek** [comment]
@jaylinski  it's random but usually happens when I try to install a package or uninstall a package with npm (un)install --save(-dev) package_name. This never happens on fresh clone and npm install of specific repo, only when npm install already run at least once.

**Comment by wangpung12** [comment]
Good

**Comment by zee92** [comment]
deleting package-lock.json then reinstalling works for me.
Window 10 
 visual studio code
node -v = v8.9.4
npm -v = 5.6.0

**Comment by zee92** [comment]
The only thing that works consistently is deleting package-lock.json to fix this but we can't keep on doing it as it contradicts purpose of package-lock.json...

**Comment by markwhitfeld** [comment]
This issue has been rolling on for months and is the primary reason why I switched to yarn.
I debugged the issue in July 2017 and found what was causing the problem I was experiencing. I added a [comment](https://github.com/npm/npm/issues/17444#issuecomment-313670843) with my findings but didn't receive any response from the dev team... although it does seem that the source code has been tweaked since then.

As mentioned in the comment, I think that the strategy for the .STAGING and .DELETE folders needs to be reviewed so that it cannot be affected by tools or access rights issues. Potentially these folders can be moved to a location that npm controls as opposed to a sub-folder of an active user's workspace. Any thoughts from the dev team?

**Comment by jaylinski** [comment]
@markwhitfeld see https://twitter.com/maybekatz/status/953402549293350913 for more info on how issues are handled by the CLI team.

Your proposal to fix this issue sounds promising!
(Although in my case the issue has to be something else, because the error happens on our Jenkins CI server.)

**Comment by TiagoSilvaPereira** [comment]
Removing package-lock.json solved for me.

**Comment by natesilva** [comment]
Hey everyone, we already know that removing `package-lock.json` works (for most people). Thank you for the reports, but they probably have enough reports of that.

The issue is that removing `package-lock.json` is an ugly ugly hack. The underlying issue needs to be found and fixed.

**Comment by jnoksoko** [comment]
Hi,

I have a clue what might be causing this. If any of your dependencies have older npm package as a dependency, then this race condition might occur.

I tried with locally installed npm version both `5.5.0` and `5.6.0`, when a third party module (auditjs) had `"npm": "^4.1.1"` in their package.json, and got this error. Upgrading this npm dependency to a latest version solved this issue for us (https://github.com/OSSIndex/auditjs/commit/1705b98dea5ea464590484318dd98363fb6006e8) for this particular package.

**Comment by adipixel** [comment]
Close all other running servers and try again!

**Comment by bloqhead** [comment]
Jumping on the bandwagon. I was trying to install `grunt-concurrent` and was running into the `ENOENT` error as seen multiple times above. My gruntfile was already prepped to use concurrent. I was otherwise ready to work.

**This is what finally fixed the problem:**

1. Remove the `package-lock.json` file (to be safe)
2. Remove the old packages with `rm -rf /node_modules/npm` (but don't remove the entire `node_modules` folder)
3. Then I was able to re-run `npm install --save-dev grunt-concurrent` without the error occurring

Now when I run my build and default commands, etc, everything is good to go.

**Handy info:**

- macOS High Sierra
- node v9.3.0
- npm v5.6.0
- VSCode

**Comment by Leexin1001** [comment]
close the vsCode,run` npm cache clean --force` , delete package-lock.json and restart computer didn't work,I solved this problem by deleting node_modules and running `npm install`,then the `npm install --save(-dev) package_name` finally work
node -v  v8.1.3
npm -v  5.6.0

**Comment by zaro** [comment]
Can confirm it's happening for me also. Never on a clean install, always on a subsequent install. Not everytime,  but most of the time. If I remove package-lock.json and rerun npm install it works.

**Comment by erikmueller** [comment]
I noticed that removing `npm4` (which is a dependency of `david` [and others i guess] and according to `npm ls dezalgo` uses it) also solves the issue.

// just saw the comment from @bloqhead ^^' that's what you get for only reading the last one.

**Comment by baoduy** [comment]
seems, delete the package-lock.json and run the install again is working for Node Js 8x as well.

**Comment by mwld** [comment]
In my case it was [Malwarebytes Anti-Ransomware](https://forums.malwarebytes.com/topic/177751-introducing-malwarebytes-anti-ransomware-beta/) preventing file access which lead to the ENOENT error. Disabling it ("Stop Protection") before running `npm install` does the job.

**Comment by john1625b** [comment]
You might get this error if your project uses a private npm registry and your are not logged into npm.
Try 

`npm login`

 and go from there

**Comment by fkleuver** [comment]
Just adding in some information which I [posted](https://github.com/npm/npm/issues/19304#issuecomment-358384940) some time ago in a related [issue](https://github.com/npm/npm/issues/19304)
## Happens only for global npm version 5.6.0
Downgrading to 5.5.1 solves it in most cases
## Will happen with any dependency on npm@4.6.1 or below
Manually upgrading an indirect dependency on npm@4.6.1 to a higher npm version, solves the issue (although this is typically a hard and cumbersome workaround in most scenarios - it's also worth asking those package authors to upgrade their npm dependency to npm@5.0.0 or up)
## Happens only when both package-lock.json and node_modules are present in the project root
Delete either (or both), and the next npm install will work again. The one after that won't.

## Clean and simple repro:
1. npm install -g npm@5.6.0
2. Create a new folder with the following package.json:
`{ "dependencies": {"npm": "^4.6.1" } }`
3. npm install -> success
4. npm install again -> fails
5. Delete node_modules and package-lock.json, then change package.json to:
`{ "dependencies": {"npm": "^5.0.0" } }`
6. npm install -> success
7. npm install again -> success
8. npm install -g npm@5.5.1
9. Repeat steps 2-7; the difference is that step 4 will succeed

**Comment by darealjulz** [comment]
`sudo npm cache clean --force` did the trick for me.

I am running an Ubuntu 16.04 (Homestead by Laravel) and using NPM 5.6.0 and Node v8.9.4.

**Comment by sscholle** [comment]
### Solution
* My explanation is that possibly the 'npm' npm module was installed locally and is conflicting...
```
rm -rf node_modules/npm
```

**Comment by retorquere** [comment]
I also have npm installed locally, but I can't prevent packages from pulling in npm.

**Comment by igordeoliveirasa** [comment]
mine was like that. however I noticed that one of my packages had a node_modules commited.. lol.

**Comment by kshivam213** [comment]
Remove node-modules folder and try once again

**Comment by alexgsoares** [comment]
> For me, functionally install with option "Do you want to use ESLint to catch probable bugs and enforce style? " equals "N".

**Comment by cancerberoSgx** [comment]
in my case I have a project with dependency projects inside that I'm referencing from the parent package.json like this: 

```sh
        "foo": "file:./innerfolder/foo"
```
removing all node_modules and package-lock.josn files recursively fixes the problem: 

```sh
find . -name "node_modules" -exec rm -rf '{}' +
find . -name "package-lock.json" -exec rm -rf '{}' +
npm install # works!
```
Workaround was removing all package-lock.json from version control :( .. Updating them all didn't made the trick. Probably Ican also solve it by executing npm install on inner projects onparent's before-install but I don't want to do that.

**Comment by alexeycrystal** [comment]
Removing of the package-lock.json file solved this problem. Thanks!

**Comment by forthgate** [comment]
Where package-lock.json stored?

**Comment by TakahiRoyte** [comment]
Removing package-lock.json insde your own project is fine, but removing it in someone else's project will deny the purpose of package-lock.json, mentioned [here](https://docs.npmjs.com/files/package-lock.json).  I recommend closing vscode, deleting ./node_modules/npm, delete cache, then if none of these works, I'd make decision to delete package-lock.json.

**Comment by tuangeek** [comment]
@mwld Malwarebytes Ransomware Protection was also causing me issues. Disabled it and npm install worked. Thanks.

**Comment by jmcollin78** [comment]
I got the problem but no package-lock.json file. Instead I've got a npm-shrinkwrap . Do you think pb may occurs when shrinkwrapping ?

**Comment by cdichiara** [comment]
Had same problem ... it was Malwarebytes' 14-day trial with real-time monitoring. Thank you @mwld!

**Comment by kornelijepetak** [comment]
Malwarebytes for me too.

**Comment by sashirajc** [comment]
I wonder if it has something to do with the latest update from Malwarebytes. Disabling malwarebytes solved this for me too

**Comment by retorquere** [comment]
I am running linux, so malwarebytes is certainly not the only reason.

**Comment by 0x08088405** [comment]
yep, malwarebytes. only figured it out because after 2 hours i remembered a few days ago it updated and i'm like hey, let's see what happens if i close it, and boom it worked.

jesus christ why

**Comment by Hoffs** [comment]
Same issue here with malwarebytes blocking npm which is fixed instantly by disabling ransomware. I also created a topic in malwarebytes forums about false positive, if anyone wants to give additional information feel free to do so: https://forums.malwarebytes.com/topic/224958-blocks-npm-from-working/

EDIT 2018-03-22: So I have received the message today from Malwarebytes staff with a new version preview and it seems they have fixed it. Installing packages worked just fine with ransomware enabled. The patch should be included in next MB update.

**Comment by emmajiugo** [comment]
Having error when trying to install `bcrypt`

My solution, I renamed the `package-lock.json` and it installed successfully

**Comment by csegyud** [comment]
Hi, I have node 9.8.0 64 bit installed along with npm 5.6.0 on a Windows 10 64 bit box. I have no package.lock.json file, I deleted node_modules folder and have run cache clear and verify several times. Tried to run earlier versions of Node (9.8.0, 8.6.0, 8.1.4, 7.10.1, switching with nvm). I uninstalled node and reinstalled. No luck. 

As a very simple test I tried to install a single package without a package.json, 

`npm i @angular/core@4.2.5`

I get 

`error enoent ENOENT: no such file or directory, rename" `

message. Please find my below complete log - a bit long. I encounter this error every time I try to "npm i" with any package.json. it stops for 20-60 sec at finalize and terminates. Only empty package.json (no deps) succeeds. :S No matter if I try to install s.g. locally or globally.

Could you guys please suggest s.g. where to find the reason? Thank you.


```
0 info it worked if it ends with ok
1 verbose cli [ 'C:\\Program Files\\nodejs\\node.exe',
1 verbose cli   'C:\\Program Files\\nodejs\\node_modules\\npm\\bin\\npm-cli.js',
1 verbose cli   'i',
1 verbose cli   '@angular/core@4.2.5' ]
2 info using npm@5.6.0
3 info using node@v9.8.0
4 verbose npm-session 3a867ea456f7ee9c
5 silly install loadCurrentTree
6 silly install readLocalPackageData
7 http fetch GET 304 https://registry.npmjs.org/@angular%2fcore 983ms (from cache)
8 silly pacote version manifest for @angular/core@4.2.5 fetched in 1089ms
9 silly install loadIdealTree
10 silly install cloneCurrentTreeToIdealTree
11 silly install loadShrinkwrap
12 silly install loadAllDepsIntoIdealTree
13 silly resolveWithNewModule @angular/core@4.2.5 checking installable status
14 http fetch GET 304 https://registry.npmjs.org/tslib 42ms (from cache)
15 silly pacote range manifest for tslib@^1.7.1 fetched in 46ms
16 silly resolveWithNewModule tslib@1.9.0 checking installable status
17 silly currentTree TESTNPM
18 silly idealTree TESTNPM
18 silly idealTree +-- @angular/core@4.2.5
18 silly idealTree `-- tslib@1.9.0
19 silly install generateActionsToTake
20 silly diffTrees action count 2
21 silly diffTrees add tslib@1.9.0
22 silly diffTrees add @angular/core@4.2.5
23 silly decomposeActions action count 16
24 silly decomposeActions fetch tslib@1.9.0
25 silly decomposeActions extract tslib@1.9.0
26 silly decomposeActions preinstall tslib@1.9.0
27 silly decomposeActions build tslib@1.9.0
28 silly decomposeActions install tslib@1.9.0
29 silly decomposeActions postinstall tslib@1.9.0
30 silly decomposeActions finalize tslib@1.9.0
31 silly decomposeActions refresh-package-json tslib@1.9.0
32 silly decomposeActions fetch @angular/core@4.2.5
33 silly decomposeActions extract @angular/core@4.2.5
34 silly decomposeActions preinstall @angular/core@4.2.5
35 silly decomposeActions build @angular/core@4.2.5
36 silly decomposeActions install @angular/core@4.2.5
37 silly decomposeActions postinstall @angular/core@4.2.5
38 silly decomposeActions finalize @angular/core@4.2.5
39 silly decomposeActions refresh-package-json @angular/core@4.2.5
40 silly install executeActions
41 silly doSerial global-install 16
42 verbose correctMkdir C:\Users\Csaba\AppData\Roaming\npm-cache\_locks correctMkdir not in flight; initializing
43 verbose lock using C:\Users\Csaba\AppData\Roaming\npm-cache\_locks\staging-9c3ac3357de07f55.lock for C:\Users\Csaba\Desktop\TESTNPM\node_modules\.staging
44 silly doParallel extract 16
45 silly extract tslib@1.9.0
46 silly pacote trying tslib@https://registry.npmjs.org/tslib/-/tslib-1.9.0.tgz by hash: sha512-f/qGG2tUkrISBlQZEjEqoZ3B2+npJjIf04H1wuAv9iA8i04Icp+61KRXxFdha22670NJopsZCIjhC3SnjPRKrQ==
47 silly extract @angular/core@4.2.5
48 silly pacote trying @angular/core@https://registry.npmjs.org/@angular/core/-/core-4.2.5.tgz by hash: sha1-YcG1iFwmIzLXN/vg9dcRUXWahGQ=
49 silly pacote tslib@https://registry.npmjs.org/tslib/-/tslib-1.9.0.tgz extracted to C:\Users\Csaba\Desktop\TESTNPM\node_modules\.staging\tslib-091c5e42 by content address 34ms
50 silly pacote @angular/core@https://registry.npmjs.org/@angular/core/-/core-4.2.5.tgz extracted to C:\Users\Csaba\Desktop\TESTNPM\node_modules\.staging\@angular\core-6d28da6c by content address 230ms
51 silly doReverseSerial unbuild 16
52 silly doSerial remove 16
53 silly doSerial move 16
54 silly doSerial finalize 16
55 silly finalize C:\Users\Csaba\Desktop\TESTNPM\node_modules\tslib
56 verbose unlock done using C:\Users\Csaba\AppData\Roaming\npm-cache\_locks\staging-9c3ac3357de07f55.lock for C:\Users\Csaba\Desktop\TESTNPM\node_modules\.staging
57 silly saveTree TESTNPM
57 silly saveTree `-- @angular/core@4.2.5
57 silly saveTree   `-- tslib@1.9.0
58 warn enoent ENOENT: no such file or directory, open 'C:\Users\Csaba\Desktop\TESTNPM\package.json'
59 verbose enoent This is related to npm not being able to find a file.
60 warn @angular/core@4.2.5 requires a peer of rxjs@^5.0.1 but none is installed. You must install peer dependencies yourself.
61 warn @angular/core@4.2.5 requires a peer of zone.js@^0.8.4 but none is installed. You must install peer dependencies yourself.
62 warn TESTNPM No description
63 warn TESTNPM No repository field.
64 warn TESTNPM No README data
65 warn TESTNPM No license field.
66 verbose stack Error: ENOENT: no such file or directory, rename 'C:\Users\Csaba\Desktop\TESTNPM\node_modules\tslib' -> 'C:\Users\Csaba\Desktop\TESTNPM\node_modules\.tslib.DELETE'
67 verbose cwd C:\Users\Csaba\Desktop\TESTNPM
68 verbose Windows_NT 10.0.16299
69 verbose argv "C:\\Program Files\\nodejs\\node.exe" "C:\\Program Files\\nodejs\\node_modules\\npm\\bin\\npm-cli.js" "i" "@angular/core@4.2.5"
70 verbose node v9.8.0
71 verbose npm  v5.6.0
72 error path C:\Users\Csaba\Desktop\TESTNPM\node_modules\tslib
73 error code ENOENT
74 error errno -4058
75 error syscall rename
76 error enoent ENOENT: no such file or directory, rename 'C:\Users\Csaba\Desktop\TESTNPM\node_modules\tslib' -> 'C:\Users\Csaba\Desktop\TESTNPM\node_modules\.tslib.DELETE'
77 error enoent This is related to npm not being able to find a file.
78 verbose exit [ -4058, true ]
```

**Comment by csegyud** [comment]
Oh, just have seen the note of @Hoffs. Earlier I thought of it on my own and added some exclusions to my MWB. Since it did not help, I dropped the idea. But now I've tried to disable ransomware protection and tadaaam, it's working like a charm. 
So, thanks @Hoffs, you saved my next day - since I already have burned a whole day on it. :) 
Waiting for MWB team to investigate the issue. Until then, we keep turning back and forth the protection. 
Cheers

**Comment by RiccoSuave** [comment]
Malwarebytes was stopping me. Thanks cdichiara.

**Comment by oshliaer** [comment]
Unfortunately, I have to remove my `package-lock.json`

Thanks  @natesilva https://github.com/npm/npm/issues/17444#issuecomment-321851320

**Comment by xerxicodex** [comment]
@mwld Dude.... malware bytes was the issue for me. Nothing was installed not even native script. spent 2 days with all of these issues. So upset that my security conscious made all this fail so hard.

**Comment by Flairieve** [comment]
I thought it was some issue with NPM or some program being ran in the background. I didn't think it would be Malwarebytes being the cause of the issue. I thought I disabled Malwarebytes but looks like I didn't. I reinstalled Node.js not thinking that Malwarebytes would be the cause but apparently it was.

If anyone is having issues with installing modules on NPM with Malwarebytes turned on, then try to either turn off the ransomware protection or exit out of Malwarebytes for right now until the Malwarebytes staff releases their newest patch!

**Comment by johndeighan** [comment]
I had this problem. It turned out to be malware bytes also. After I disabled it, I could install packages with no problem.

**Comment by bwobbones** [comment]
I don't think this is a malwarebytes issue, at least its not **just** a malwarebytes issue.  I don't have malwarebytes, this happens to me within a corporate network where the disk speed is very slow.  I have turned off all virus scanning and still can't get an npm install to work.  Maybe the link is that all new files are being scanned and slowing down the I/O, exposing the issue.

I have done a little tracing through the npm code, through finalize.js specifically, working on the theory that there is a race condition between the time that a file is deleted from the .staging directory and when it is moved back to the node_modules file.  No result yet, will report back if I find anything.

**Comment by GsHeri** [comment]
malwarebytes for me too

**Comment by sjohnsonaz** [comment]
Same problem, removing malwarebytes fixed it.

**Comment by solinad21** [comment]
malwarebytes for me too

**Comment by biship** [comment]
malwarebytes for me too

**Comment by retorquere** [comment]
OK people, we now know that malwarebytes is an issue. Everyone in this thread gets mail updates whenever someone adds a comment -- you can add status icons to existing comments, and people won't get email for those. I don't use Windows so I don't use MWB, and I'm *also* seeing the issue, so I'm lurking here hoping to find a solution for non-windows systems.

**Comment by Bessonov** [comment]
> OK people, we now know that malwarebytes is an issue.

I don't have malwarebytes on my linux and experienced this issue too.

And please, no more "malwarebytes for me too".

**Comment by vincentntang** [comment]
Had malware bytes this was my problem as well, for some reason malware bytes had also caused me general windows issues such as being unable to rename folders

**Comment by nicbuitr** [comment]
What solved it for me:

- Disabled "Windows Defender - Real Time Protection".
- Ran "npm install" again.
- Worked.

After reading all the comments about malwarebytes, I tried that, as it seems like malware protection prevents the correct installation of the node_modules on the AppData folder for security reasons.

**Comment by jakeNiemiec** [comment]
@nicbuitr Windows Defender? 

Are you telling me that npm won't work on a clean windows install without disabling safeguards? You might want to create a separate issue for that.

**Comment by TehShrike** [comment]
I don't believe there's a need for another issue.  This issue is sufficient and serious enough to pay attention to as it is, we shouldn't have to create duplicates.

**Comment by nicbuitr** [comment]
@jakeNiemiec well, not actually, npm worked well without any workaround for most of my projects, it just showed this error for one of them in particular that I did a year ago and need to work again with, so I cloned it and when trying running "npm install" it showed the error reported, I tried everything but it only worked when I disabled the "Windows Defender - Real Time Protection".

So it seems like one of the project's npm dependencies was not being allowed to install on that folder by the malware protection, couldn't spot which specifically, however, here I attach the debug log for if it can be of any use:

[2018-04-04T11_19_56_356Z-debug.log](https://github.com/npm/npm/files/1876756/2018-04-04T11_19_56_356Z-debug.log)

**Comment by kevinhaas** [comment]
Randomly started getting this tonight, turned out to be Malwarebytes for me as well. I've had it running for a long time while using npm and never had this pop up until now. Not sure what triggered it.

**Comment by km9554** [comment]
ugh, malwarebytes. same problem.

**Comment by retorquere** [comment]
For any newcomers to this thread, *yes, we know*, MWB causes this. Please star existing comments to show your malcontent, or to flag that you too use MWB and you're hitting this issue. Just saying "hey, me too!" doesn't add any extra diagnostic information at this point.

**Comment by ghost** [comment]
For me it was Malwarebytes the whole f*** time i had to uninstall it ,and it did work smoothly lol.

**Comment by retorquere** [comment]
And another spurious "hey MWB for me too lol". Did you not read the thread @flushedfriendd? **we know**. Christ.

**Comment by kevinhaas** [comment]
You should probably close it if you care that much. Seems like you just enjoy bitching at folks, though.

**Comment by tobiasmuecksch** [comment]
I'm sorry, but what is MWB?

**Comment by jakeNiemiec** [comment]
@kevinhaas MWB (MalWareBytes) is a symptom to a deeper problem, the larger issue has to do with "no such file or directory" on *nix systems **or** that the actual error message is much too generic to troubleshoot. I am sure the npm team (more specifically @iarna) is aware of this.

**While I am glad your problem was resolved**, those of us who are getting this without MWB installed don't appreciate hundreds of "**Me too, thanks**." showing up in our inboxes.

**Comment by ghost** [comment]
@retorquere  IT'S JUST A COMMENT calm down jesus christ.

**Comment by nickserv** [comment]
Please lock this thread.

**Comment by solinad21** [comment]
Hello, thinking badly (related to mwb), could there be the possibility that some dependency of our projects could infiltrate unsafe code or unsafe actions or traces of "something" unsafe? I mention this because at least I try to have a hardaware superior to that of a normal user and that is very much desired, for example, by miners.

Excuse me if this message is not appropriate to the subject.

**Comment by scottdear** [comment]
If I remove package-lock.json it works for me.  I am on *nix, have killed all my node processes and I am not running VSCode.  It appears there is some issue with how it uses package-lock.json.

**Comment by callenj** [comment]
MWB 3.4.5 solves this issue.  (or did for me).  You can find the discussion and download link here:
https://forums.malwarebytes.com/topic/224958-blocks-npm-from-working/

**Comment by chadahahn** [comment]
Hi everyone, spent a couple of hours before turning off MalwareBytes Ransomware Protection, and finally got things to work.

**Comment by mathieulj** [comment]
I have just encountered this very issue. I have no node processes active and I dont even have vscode installed. Removing the package-lock.json solves the issue for me. In my case the errors only seems to occur in the dependencies of a 'file://../other_package' module so maybe it's related to the symlink?

Example:
```
> git checkout HEAD -- package-lock.json
> npm install
...
ENOENT: no such file or directory, rename '/opt/other_package/node_modules/commander' -> '/opt/other_package/node_modules/.commander.DELETE'
...
> rm package-lock.json
> npm install
...
added 140 packages from 90 contributors in 9.629s
```

Edit:
Debian Stretch x64, node v8.11.1 & npm 5.8.0

**Comment by nicbuitr** [comment]
Last night I had this problem again but turning off Windows Defender didn't work as last time, so what I ended up doing was copying the "node_modules" folder from other project to the one I was trying to setup, then "npm install" and it worked.

The downside of this, is that I don't know if "npm install" removes the extra modules that are not required by the new project, therefore having extra module folders that are not being used.

**Comment by HikmetTenis** [comment]
Im getting ,DELETE issue in Ubuntu 16.04 with npm 3,5,2.. I got no VSCODE, no Malware, no Defender, not sure node-gyp would cause any sharing issue, if I run npm install individually for that package, it works, it is my workaround anyways but I dont like workarounds.

**Comment by GCSBOSS** [comment]
For me the problem was solved ~apparently~ with `npm uninstall ...` and then `npm install ...`.

**Comment by klianlopez** [comment]
I was installing using bash in powershell. I tried switching to normal cmd and it works

**Comment by kcrwfrd** [comment]
Just experienced this on MacOS 10.13.4 with npm 5.6.0, and no other instances of node running. Removing the `package-lock.json` file got it working.

**Comment by linzx1993** [comment]
![image](https://user-images.githubusercontent.com/18568545/40586923-3d047dfe-61fb-11e8-8690-368cf78b60a4.png)

First of all i use private proxy npm。
 i have same error in node@8.10.0
but good when i reset node version to @6.11.3

**Comment by alexgorbatchev** [comment]
Just experienced this issue with npm 6.x. Closing VScode didn’t work for me. **I had to delete `package-lock.json` file, then everything started working as expected.**

**Comment by miensol** [comment]
Same problem here. Removing `package-lock.json` makes the problem go away. However, **this is by no means a solution nor a workaround**.

**Comment by itsmepetrov** [comment]
Same here

**Comment by Hychhayrith** [comment]
Hello everyone, for this problem just run npm update after that you can install module as usual

**Comment by zkat** [comment]
This looks like it's pretty much resolved for most folks.

To be clear: If you're experiencing this:

1. Make sure you have the latest npm (`npm install -g npm`).
2. Add an exception to your antivirus to ignore the `node_modules` folder in your project.
3. `$ rm -rf node_modules package-lock.json`
4. `$ npm install`

And you should be back in business.

I'm going to both close and lock this, since it's resolved. If you're still experiencing this and you have a reliable repro, and none of the above steps fixed it for you please file a new issue [in the npm.community forum](https://npm.community) with any necessary repro steps so that we can look into it, in case there's still a missing bit. Cheers!
